src/RA_Dlg_MemBookmark.cpp

```diff
@@ -704,50 +704,43 @@ void Dlg_MemBookmark::ExportJSON()
     defaultDir.erase(0, 2); // Removes the characters (".\\")
     defaultDir = g_sHomeDir + defaultDir;
 
-    IFileSaveDialog* pDlg = nullptr;
-    HRESULT hr = CoCreateInstance(CLSID_FileSaveDialog, nullptr, CLSCTX_ALL, IID_IFileSaveDialog, reinterpret_cast<void**>(&pDlg));
-    if (hr == S_OK)
+    CComPtr<IFileSaveDialog> pDlg;
+
+    if (auto hr = CoCreateInstance(CLSID_FileSaveDialog, nullptr, CLSCTX_ALL, IID_IFileSaveDialog,
+        reinterpret_cast<void**>(&pDlg)); SUCCEEDED(hr))
     {
-        hr = pDlg->SetFileTypes(ARRAYSIZE(c_rgFileTypes), c_rgFileTypes);
-        if (hr == S_OK)
+        if (SUCCEEDED(hr = pDlg->SetFileTypes(c_rgFileTypes.size(), &c_rgFileTypes.front())))
         {
-            char defaultFileName[512];
-            sprintf_s(defaultFileName, 512, "%s-Bookmarks.txt", std::to_string(g_pCurrentGameData->GetGameID()).c_str());
-            hr = pDlg->SetFileName(Widen(defaultFileName).c_str());
-            if (hr == S_OK)
+            std::ostringstream oss;
+            oss << g_pCurrentGameData->GetGameID() << "-Bookmarks.txt";
+
+            if (auto defaultFileName{ oss.str() }; SUCCEEDED(hr = pDlg->SetFileName(Widen(defaultFileName).c_str())))
             {
-                PIDLIST_ABSOLUTE pidl;
-                hr = SHParseDisplayName(Widen(defaultDir).c_str(), nullptr, &pidl, SFGAO_FOLDER, 0);
-                if (hr == S_OK)
+                if (PIDLIST_ABSOLUTE pidl{ nullptr }; SUCCEEDED(hr = SHParseDisplayName(Widen(defaultDir).c_str(), nullptr, &pidl, SFGAO_FOLDER, nullptr)))
                 {
-                    IShellItem* pItem = nullptr;
-                    SHCreateShellItem(nullptr, nullptr, pidl, &pItem);
-                    hr = pDlg->SetDefaultFolder(pItem);
-                    if (hr == S_OK)
+                    CComPtr<IShellItem> pItem;
+
+                    if (SHCreateShellItem(nullptr, nullptr, pidl, &pItem); SUCCEEDED(hr = pDlg->SetDefaultFolder(pItem)))
                     {
-                        pDlg->SetDefaultExtension(L"txt");
-                        hr = pDlg->Show(nullptr);
-                        if (hr == S_OK)
+                        if (pDlg->SetDefaultExtension(L"txt"); SUCCEEDED(hr = pDlg->Show(nullptr)))
                         {
-
-                            hr = pDlg->GetResult(&pItem);
-                            if (hr == S_OK)
+                            if (SUCCEEDED(hr = pDlg->GetResult(&pItem.p)))
```

Comment:
https://docs.microsoft.com/en-us/cpp/atl/reference/ccomptrbase-class#operator_amp
It seems the `CComPtr `needs to be `null `to use the `address-of operator` on it. 
Since it was already used the framework thinks we have buggy code, if
we truly need the address then we need store the address explicit. Which appears to be the case.
We would never have known with just the raw interface pointer
		   
Straight from` atlcomcli.h: CComPtrBase T** operator&()`
The `assert `on `operator&` usually indicates a bug. 
If this is really what is needed, however, take the address of the `p `member explicitly.